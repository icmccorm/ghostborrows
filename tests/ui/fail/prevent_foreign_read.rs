use ghostborrows::*;

fn main() {
    Value::new(1).borrow_mut(|r1, token1| {
        let activated = r1.activate(token1);
        activated.borrow_mut(|_, _| {
            print!("{}", *r1);
        });
    });
}