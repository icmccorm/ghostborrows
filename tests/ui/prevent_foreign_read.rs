use ghost_borrows::*;

fn main() {
    OwnedValue::new(1).borrow_mut(|r1, token1| {
        let activated = r1.activate(token1);
        activated.borrow_mut(|_, _| {
            activated.read(|r1m| print!("{}", *r1m) );
        });
    });
}