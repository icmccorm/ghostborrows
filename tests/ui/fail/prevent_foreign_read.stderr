error[E0382]: borrow of moved value: `r1`
 --> tests/ui/fail/prevent_foreign_read.rs:6:30
  |
4 |     Value::new(1).borrow_mut(|r1, token1| {
  |                               -- move occurs because `r1` has type `RefReserved<'_, i32>`, which does not implement the `Copy` trait
5 |         let activated = r1.activate(token1);
  |                            ---------------- `r1` moved due to this method call
6 |         activated.borrow_mut(|_, _| {
  |                              ^^^^^^ value borrowed here after move
7 |             print!("{}", *r1);
  |                           -- borrow occurs due to use in closure
  |
note: `RefReserved::<'tag, T>::activate` takes ownership of the receiver `self`, which moves `r1`
 --> src/refs.rs
  |
  |     pub fn activate(self, _token: Token<'tag, T>) -> RefMut<'tag, T> {
  |                     ^^^^
